### PR TITLE
RATIS-2068. Avoid logging raw StateMachine data body.

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

See [RATIS-2068](https://issues.apache.org/jira/browse/RATIS-2068).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-2068

## How was this patch tested?

CI

### CODE REVIEW

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcServerProtocolService.java [hunk 1]
```diff
@@ -316,7 +316,7 @@ boolean isHeartbeat(AppendEntriesRequestProto request) {
 
       @Override
       String requestToString(AppendEntriesRequestProto request) {
-        return ServerStringUtils.toAppendEntriesRequestString(request);
+        return ServerStringUtils.toAppendEntriesRequestString(request, e -> "");
```
Comment (szetszwo) [inline ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcServerProtocolService.java:319]: Let's pass `null`.  The code currently can handle `null`.

BTW, could you also update `stateMachineLogEntryProtoToString(..)` to print only sizes and 
`toLogEntryString(..)` to use `stateMachineLogEntryProtoToString(..)`?
```java
@@ -70,7 +70,11 @@ public final class LogProtoUtils {
   }
 
   static String stateMachineLogEntryProtoToString(StateMachineLogEntryProto p) {
-    return "logData:" + p.getLogData() + ", stateMachineEntry:" + p.getType() + ":" + p.getStateMachineEntry();
+    final StateMachineEntryProto stateMachineEntry = p.getStateMachineEntry();
+    return p.getType()
+        + ": logData.size=" + p.getLogData().size()
+        + ", stateMachineData.size=" + stateMachineEntry.getStateMachineData().size()
+        + ", logEntryProtoSerializedSize=" + stateMachineEntry.getLogEntryProtoSerializedSize();
   } 
```

```java
@@ -46,9 +45,10 @@ public final class LogProtoUtils {
     }
     final String s;
     if (entry.hasStateMachineLogEntry()) {
-      s = ", " + Optional.ofNullable(function)
-          .orElseGet(() -> proto -> "" + ClientInvocationId.valueOf(proto))
-          .apply(entry.getStateMachineLogEntry());
+      if (function == null) {
+        function = LogProtoUtils::stateMachineLogEntryProtoToString;
+      }
+      s = ", " + function.apply(entry.getStateMachineLogEntry());
     } else if (entry.hasMetadataEntry()) {
       final MetadataProto metadata = entry.getMetadataEntry();
       s = "(c:" + metadata.getCommitIndex() + ")";
```

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java [hunk 2]
```diff
@@ -83,10 +85,12 @@ public abstract class RaftLogBase implements RaftLog {
   private final long purgePreservation;
 
   private final AtomicReference<LogEntryProto> lastMetadataEntry = new AtomicReference<>();
+  private final Function<StateMachineLogEntryProto, String> stateMachineToString;
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java:88]: Let don't add `StateMachine` dependency `RaftLogBase` . The subclasses can override `toLogEntryString(..)`.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java [hunk 3]
```diff
@@ -199,10 +208,10 @@ private long appendImpl(long term, TransactionContext operation) throws StateMac
 
       appendEntry(operation.wrap(e), operation).whenComplete((returned, t) -> {
         if (t != null) {
-          LOG.error(name + ": Failed to write log entry " + LogProtoUtils.toLogEntryString(e), t);
+          LOG.error(name + ": Failed to write log entry " + toString(e), t);
         } else if (returned != nextIndex) {
           LOG.error("{}: Indices mismatched: returned index={} but nextIndex={} for log entry {}",
-              name, returned, nextIndex, LogProtoUtils.toLogEntryString(e));
+              name, returned, nextIndex, toString(e));
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java:214]: Use `this.toLogEntryString(..)`.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java [hunk 4]
```diff
@@ -532,11 +541,8 @@ private void discardData() {
...
     @Override
     public String toString() {
-      return toLogEntryString(logEntry);
+      return RaftLogBase.this.toString(logEntry);
     }
   }
 
-  public String toLogEntryString(LogEntryProto logEntry) {
-    return LogProtoUtils.toLogEntryString(logEntry);
-  }
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java:541]: Let's keep this method.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java [hunk 5]
```diff
@@ -345,8 +347,7 @@ public ReferenceCountedObject<EntryWithData> retainEntryWithData(long index) thr
       }
       return future != null? newEntryWithData(entryRef, future): newEntryWithData(entryRef);
     } catch (Exception e) {
-      final String err = getName() + ": Failed readStateMachineData for " +
-          LogProtoUtils.toLogEntryString(entry);
+      final String err = getName() + ": Failed readStateMachineData for " + toString(entry);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java:350]: Use `this.toLogEntryString(..)`.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java [hunk 6]
```diff
@@ -467,7 +468,7 @@ protected CompletableFuture<Long> appendEntryImpl(ReferenceCountedObject<LogEntr
       }
       return write.getFuture().whenComplete((clientReply, exception) -> appendEntryTimerContext.stop());
     } catch (Exception e) {
-      LOG.error("{}: Failed to append {}", getName(), LogProtoUtils.toLogEntryString(entry), e);
+      LOG.error("{}: Failed to append {}", getName(), toString(entry), e);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java:471]: Use `this.toLogEntryString(..)`.

**Review by szetszwo — COMMENTED** [review]
@duongkame , thanks a lot for fixing this!  Please see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13068605/1074_review.patch

**Comment by duongkame** [comment]
> @duongkame , thanks a lot for fixing this! Please see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13068605/1074_review.patch

Thanks for the quick review @szetszwo . Updated the PR.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**